feat(CenterXYDownloads)

AlyAsh opened the pull request:
Asana task: https://app.asana.com/0/1203496001148058/1204293328884281/f

- Adds the ability to download CenterXY Collections to `createCollectionDownload.py`. 
- Adds the input `cache_time` for triggering a new download if the latest one is older than the cache_time.
- Some refactoring and log/print changes.

AlyAsh committed:
feat(CenterXYCollectionDownload)

adds the feature to download CenterXY Collections, to include a cache_time to trigger a new download based on the difference in time between now and the most recent download

AlyAsh committed:
Update README.md

AlyAsh commented:
@rrydv I noticed a bug with using the old ChainListDownload endpoint where it would say that some ChainLists don't exist / have no record, e.g. 245082, so I changed it to the generic Downloads endpoint. This bug is present in the Alteryx Plugin, so if this is approved, can I make the change to Tools directly to main?

AlyAsh committed:
fix removed my API key

AlyAsh committed:
fix reset collection_id to 0

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -2,75 +2,189 @@
+    
+    return records_url, check_url, new_download_url, url_params
+
+def retrieve_new_download(create_date, cache_time:int):
```
I think this method needs to be renamed, maybe something like is_download_stale?

AlyAsh replied:
Renamed to `is_download_stale`

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -2,75 +2,189 @@
+    if cache_time < 0:
+        raise ValueError("Please enter a valid duration for cache_time that is an int > 0.")
+
+def get_collection_urls(collection_id:str, collection_type:str):
```
What's the benefit of this living as a separate function? It seems like it just sets the values conditionally.

AlyAsh replied:
I was thinking to declutter `download_collection()` and have it just focus on the actions related to downloading a collection as the name suggests, but the assignments are readable in either case.

After thinking about it more, I don't think we'll ever use `get_collection_urls()` for any other purpose than in `download_collection()`, and it's pretty obvious where it is if we ever need to change it, so I'll revert this change.

Resolved!

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -2,75 +2,189 @@
+    r_body = json.loads(r.text)
+    return r_body['Id']
+
+def generate_download(collection_download_id:int, headers:str):
```
I find this method name a bit confusing as well, we already generated a download in the create_new_download. We're only checking the download status here.

AlyAsh replied:
Renamed to `get_download_link` to better describe what the method does

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -2,75 +2,189 @@
+
+    # IF A RECORD DOES NOT EXIST, CHECK IF THE COLLECTION EXISTS WITH 0 DOWNLOADS AND CREATE A NEW ONE;
+    # OTHERWISE, THERE IS NO DOWNLOAD TO BE MADE
+    if not records:
```
this section needs to be revised, since this file doesn't deal with generating a new collection. This file is only meant for downloading existing collections. As such, functionality that deals with generating a new record etc. needs to be deleted here. You can use it to modify https://github.com/ChainXY/ChainXY_API/blob/main/python/createCollectionAndDownload.py that deals with creating a new collection.

rrydv replied:
if you pursue this, I'd suggest splitting methods for generating ChainLists and SiteLists, the underlying entities are different enough that mixing them would just make the record creation very messy.

AlyAsh replied:
Comment should be "If a download record doesn't exist, check if the collection exists with 0 downloads then create a new download." I'll change the comment but keep the ability to create a new download if the most recent download is stale

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -2,75 +2,189 @@
+    result = json.loads(r.text)
+    if 'Message' in result:
+        if 'record does not exist' in result['Message'].lower():
+            print("No collection found. Please ensure you have the correct collection ID and collection Type.")
```
this should raise a ValueError in this context (i.e. error out if collection doesn't exist)

AlyAsh replied:
Changed to ValueError to make the program simply error out if the collection doesn't exist

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -80,7 +194,7 @@ def download_file(url:str, output_file:str):
                 #if chunk: 
                 f.write(chunk)
 
-    print(f"Saved {url}\nto\n{output_file}")
```
I think we should keep the url in the print statement here, otherwise you can't copy+paste to try to download the file (without re-running the whole script and printing/logging the url of the generated download.

AlyAsh replied:
I moved the URL print to the end of `download_collection()` because it would print twice and fill the console, but I can put it back here

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -2,75 +2,189 @@
+from datetime import datetime
 
-def check_api_key(cxy_api_key):
+def request_api(url:str, headers:str, method='GET', params=None, data={}):
```
I'm not sure if it makes sense to just move the 

```
    headers = {
        'x-apikey': cxy_api_key,
        'x-Application': 'Python API Call',
        'content-type': 'application/json'
        }
```
from donwload_collection to this method, and instead accept cxy_api_key as input? I don't see why we'd want to mess around with headers outside of the method calling the api

AlyAsh replied:
True. I was going back and forth between the headers of the API key but you're right, it makes more sense to have it in the method that deals with the requests. Resolved!

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -2,75 +2,189 @@
+        url_params = {
+            "format": "ZIP_CSV"
+        }
+    
```
should there be another else here that raises ValueError if the collection type doesn't match the accepted collection_types?

AlyAsh replied:
The check is in `validate_inputs()`:

```
if not isinstance(collection_type, str) or collection_type.lower() not in ['chain', 'center']:
    raise ValueError(f"Collection type '{collection_type}' must be one of ['chain', 'center'].")
```

rrydv reviewed (changes requested):
please address the feedback/questions in the review. For this file, we don't really want to  create new collections, it's really only meant as a lightweight tool to download your pre-configured collections. It's great that you already implemented handling of cases where a collection doesn't exist, but it needs to be handled in a different file (see comments)

AlyAsh committed:
addresses changes requested

rrydv committed:
refactor(createCollectionDownload) remove unused function

rrydv commented on python/createCollectionDownload.py:
```diff
@@ -1,76 +1,209 @@
+    ## GET THE APPROPRIATE API REQUEST URLS FOR COLLECTION TYPE
+    records_url = check_url = new_download_url = ''
+    if collection_type == 'chain':
+        records_url = f'https://location.chainxy.com/api/Downloads?fields=Id,ChainList,Label,User,ReportType,CreateDate,DataDate,Status,Format,Count,Link,DisplayValue,&Query=%7B%22ChainListId%22:{collection_id}%7D&OrderBy=-CreateDate'
```
we need to make sure that we're filtering out other report types when we search for downloads here, when I ran this it picked up a ChangesOverTime report, even though it's only meant to pick up ChainLocationReports. See ReportType enum in the backend.

rrydv committed:
feat(createCollectionDownload) let users overwrite report parameters

rrydv reviewed (changes requested):
please address the feedback in the comment, also I wasn't able to download a center collection 397. it errors out.

AlyAsh committed:
fix params not passed to request_api & added reportType to records url

AlyAsh commented:
> ... also I wasn't able to download a center collection 397. it errors out.

The error was caused by the 'params' parameter being changed to 'json' on line 16:
`r = requests.request(url=url, method=method, headers=headers, json=params, data=data)` 
and since the SiteList Controller requires the ?format=ZIP_CSV parameter in the endpoint, it errored out since it was cut off. Given that we allow custom params, I wonder if it should just be hardcoded in.

It should be fixed in the latest commit, along with the updated request URLs filtering by report types (0 - chain, 4 - center), status, and format. I removed the fields param as well since it's redundant and I get the same response with or without.

rrydv committed:
fix(createCollectionDownload) remove filters for format/status

filters are unnecessary, as clients might request formats other than csv, and we want to see report with failed status

rrydv committed:
refactor(createCollectionDownload) use pythonic boolean
